File: Editor/EasyRPG.Editor.Serializer/RM2K/LMU/LmuEventPage.cs
using EasyRPG.Editor.RPG;
using EasyRPG.Editor.Serializer.RM2K.Common;

namespace EasyRPG.Editor.Serializer.RM2K.LMU
{
	public partial class LmuReader
	{
		#region BehaviorMethods
		private EventPage ReadEventPage()
		{
			var eventpage = new EventPage
								{
									Id = stream.ReadCompressedInt()
								};

			while (!stream.EndOfFile())
			{
				var chunkId = (ChunkEventPage)stream.ReadCompressedInt();
				if (chunkId == (int)ChunkData.End)
				{
					break;
				}

				var chunkLenght = stream.ReadCompressedInt();
				if (chunkLenght == 0)
				{
					continue;
				}

				switch (chunkId)
				{
					case ChunkEventPage.Condition:
						eventpage.Condition = ReadEventPageCondition();
						break;
					case ChunkEventPage.CharacterName:
						eventpage.CharacterName = stream.ReadString(chunkLenght);
						break;
					case ChunkEventPage.CharacterIndex:
						eventpage.CharacterIndex = stream.ReadCompressedInt();
						break;
					case ChunkEventPage.CharacterDirection:
						eventpage.CharacterDirection = stream.ReadCompressedInt();
						break;
					case ChunkEventPage.CharacterPattern:
						eventpage.CharacterPattern = stream.ReadCompressedInt();
						break;
					case ChunkEventPage.Translucent:
						eventpage.Translucent = stream.ReadCompressedBool();
						break;
					case ChunkEventPage.MoveType:
						eventpage.MoveType = stream.ReadCompressedInt();
						break;
					case ChunkEventPage.MoveFrequency:
						eventpage.MoveFrequency = stream.ReadCompressedInt();
						break;
					case ChunkEventPage.Trigger:
						eventpage.Trigger = stream.ReadCompressedInt();
						break;
					case ChunkEventPage.PriorityType:
						eventpage.PriorityType = stream.ReadCompressedInt();
						break;
					case ChunkEventPage.Overlap:
						eventpage.Overlap = stream.ReadCompressedBool();
						break;
					case ChunkEventPage.AnimationType:
						eventpage.AnimationType = stream.ReadCompressedInt();
						break;
					case ChunkEventPage.MoveSpeed:
						eventpage.MoveSpeed = stream.ReadCompressedInt();
						break;
					case ChunkEventPage.MoveRoute:
						eventpage.MoveRoute = ReadMoveRoute();
						break;
					case ChunkEventPage.EventCommandsSize:
						stream.ReadCompressedInt();
						break;
					case ChunkEventPage.EventCommands:
						for (;;)
						{
							if (stream.PeekChar() == 0)
							{
								stream.Skip(4);
								break;
							}
							eventpage.EventCommands.Add(EventCommandReader.ReadEventCommand(stream));
						}
						break;
					default:
						stream.Skip(chunkLenght);
						break;
				}
			}
			return eventpage;
		}
		#endregion
	}
}

File: readers/include/detail/lmu_eventpage.cpp
/////////////////////////////////////////////////////////////////////////////
// This file is part of EasyRPG.
//
// EasyRPG is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// EasyRPG is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with EasyRPG Player. If not, see <http://www.gnu.org/licenses/>.
/////////////////////////////////////////////////////////////////////////////

////////////////////////////////////////////////////////////
// Headers
////////////////////////////////////////////////////////////
#include "lmu_reader.h"
#include "lmu_chunks.h"

////////////////////////////////////////////////////////////
/// Read Event Page
////////////////////////////////////////////////////////////
EASYRPG_STRUCT_ID_READER(EventPage, WithID)

EASYRPG_STRUCT_NAME(EventPage)

#define EASYRPG_CHUNK_SUFFIX LMU_Reader
#define EASYRPG_CURRENT_STRUCT EventPage

EASYRPG_STRUCT_FIELD_MAP_BEGIN(EventPage)
	, EASYRPG_TYPED_FIELD_PAIR(RPG::EventPageCondition, condition, ('cond')('itio')('n'))
	, EASYRPG_TYPED_FIELD_PAIR(std::string, character_name, ('char')('acte')('r_na')('me'))
	, EASYRPG_TYPED_FIELD_PAIR(int, character_index, ('char')('acte')('r_in')('dex'))
	, EASYRPG_TYPED_FIELD_PAIR(int, character_direction, ('char')('acte')('r_di')('rect')('ion'))
	, EASYRPG_TYPED_FIELD_PAIR(int, character_pattern, ('char')('acte')('r_pa')('tter')('n'))
	, EASYRPG_TYPED_FIELD_PAIR(bool, translucent, ('tran')('sluc')('ent'))
	, EASYRPG_TYPED_FIELD_PAIR(int, move_type, ('move')('_typ')('e'))
	, EASYRPG_TYPED_FIELD_PAIR(int, move_frequency, ('move')('_fre')('quen')('cy'))
	, EASYRPG_TYPED_FIELD_PAIR(int, trigger, ('trig')('ger'))
	, EASYRPG_TYPED_FIELD_PAIR(int, priority_type, ('prio')('rity')('_typ')('e'))
	, EASYRPG_TYPED_FIELD_PAIR(bool, overlap, ('over')('lap'))
	, EASYRPG_TYPED_FIELD_PAIR(int, animation_type, ('anim')('atio')('n_ty')('pe'))
	, EASYRPG_TYPED_FIELD_PAIR(int, move_speed, ('move')('_spe')('ed'))
	, EASYRPG_TYPED_FIELD_PAIR(RPG::MoveRoute, move_route, ('move')('_rou')('te'))
	, EASYRPG_SIZE_FIELD_PAIR(RPG::EventCommand, event_commands, ('even')('t_co')('mman')('ds'))
	, EASYRPG_TYPED_FIELD_PAIR(std::vector<RPG::EventCommand>, event_commands, ('even')('t_co')('mman')('ds'))
EASYRPG_STRUCT_FIELD_MAP_END()

#undef EASYRPG_CURRENT_STRUCT
#undef EASYRPG_CHUNK_SUFFIX
